SingularityExecutor/src/main/java/com/hubspot/singularity/executor/task/SingularityExecutorTaskProcessCallable.java

```diff
@@ -115,4 +131,93 @@ private boolean runHealthcheck(Process process) {
       return false;
     }
   }
+
+  private boolean runHealthcheckV2(Process process, HealthcheckOptions maybeOptions) {
+    String taskAppDirectory = task.getTaskDefinition().getTaskAppDirectory();
+    File configPath = Paths.get(taskAppDirectory, HEALTHCHECK_CONFIG_FILE_PATH).toFile();
+    LOG.debug("Getting healthcheck config file path at {}", configPath);
+
+    HealthchecksV2 healthchecks;
+    try {
+      Retryer<Boolean> retryer = RetryerBuilder.<Boolean>newBuilder()
+          .retryIfResult(x -> !x)
+          .withWaitStrategy(WaitStrategies.fixedWait(500, TimeUnit.MILLISECONDS))
+          .withStopStrategy(StopStrategies.stopAfterAttempt(20))
+          .build();
+      retryer.call(() -> configPath.exists());
+      healthchecks = objectMapper.readValue(configPath, HealthchecksV2.class);
+    } catch (IOException e) {
+      executorUtils.sendStatusUpdate(task.getDriver(), task.getTaskInfo().getTaskId(), TaskState.TASK_FAILED, String.format("Failed to run healthchecks. Failed to read healthcheck config file. %s", e.getCause()), task.getLog());
+      return false;
+    } catch (ExecutionException | RetryException e) {
+      executorUtils.sendStatusUpdate(task.getDriver(), task.getTaskInfo().getTaskId(), TaskState.TASK_FAILED, String.format("Failed to run healthchecks. Failed to find healthcheck config file. %s", e.getCause()), task.getLog());
+      return false;
+    }
+
+    // Start all the futures, ordered by max possibe lifespan
+    List<CompletableFuture<HealthcheckResult>> futures = healthchecks.getHealthchecks().stream()
+        .filter(HealthcheckConfig::isEnabled)
+        .sorted(Comparator.comparingInt(config -> config.getTimeout()))
+        .map(healthcheckConfig -> startHealthcheck(process, healthcheckConfig))
+        .collect(Collectors.toList());
+
+    List<HealthcheckResult> results = futures.stream().map(CompletableFuture::join).collect(Collectors.toList());
+
+    for (HealthcheckResult result : results) {
+      HealthCheckStatus status = result.getStatus();
+      if (status != HealthCheckStatus.PASSED) {
+        executorUtils.sendStatusUpdate(task.getDriver(), task.getTaskInfo().getTaskId(), TaskState.TASK_FAILED, String.format("Healthcheck failed: %s", result.getMessage()), task.getLog());
+        return false;
+      }
+    }
+    executorUtils.sendStatusUpdate(task.getDriver(), task.getTaskInfo().getTaskId(), Protos.TaskState.TASK_RUNNING, String.format("Task running process %s.", getCurrentProcessToString()), task.getLog());
+    return true;
+  }
+
+  private CompletableFuture<HealthcheckResult> startHealthcheck(Process process, HealthcheckConfig healthcheckConfig) {
+    Retryer<HealthcheckResult> retryer = buildHealthcheckRetryer(healthcheckConfig);
+    return CompletableFuture.supplyAsync( () -> {
```

Comment:
We should make sure to use an ExecutorService with this. Ideally shut it down right after as well